### PR TITLE
Change TentacleClient.ExecuteScript to use new ExecuteScriptCommand type

### PR DESCRIPTION
**Pull request by APErebus**
# Background

Currently, `TentacleClient.ExecuteScript()`  takes a `StartScriptCommandV3Alpha`, which ties the contract for executing a script to a specific script service version (in this case V3Alpha).

This means that the contract of TentacleClient is tied to the latest version of the Script Service, which then means over the halibut-wire contracts are being propagated throughout Server's codebase. There is an additional problem then for the upcoming `KubernetesScriptServiceV1Alpha` as it will have it's own `StartScriptCommand` and `ScriptServiceV3Alpha` will be going away.

This PR introduces a new abstract `ExecuteScriptCommand` and two concrete implementations, `ExecuteShellScriptCommand` and `ExecuteKubernetesScriptCommand` types which are now the public contract for `TentacleClient.ExecuteScript()`.

# Results

Introduced the three new types and updated the TentacleClient and script orchestration to use the new types. Also included the builders I developed in https://github.com/OctopusDeploy/OctopusDeploy/pull/23506 into the TentacleClient so we have consistent builders for consumption

Almost all the changes in the tests are moving away from the `LatestStartScriptCommandBuilder` to a new `TestExecuteShellScriptCommandBuilder`. The `StartScriptCommandV3Alpha` build remains as it's required for the `ScriptServiceV3AlphaFixture`

Shortcut story: [sc-73757]

# How to review this PR

Quality :heavy_check_mark:

# Pre-requisites

- [ ] I have read [How we use GitHub Issues](https://github.com/OctopusDeploy/Issues/blob/main/docs/CONTRIBUTING.internal.md) for help deciding when and where it's appropriate to make an issue.
- [x] I have considered informing or consulting the right people, according to the [ownership map](https://whimsical.com/ownership-map-NzbiD4HJyvhC9jNJNfS6TG).
- [x] I have considered appropriate testing for my change.

### CODE REVIEW

**Comment by jimmyp** [comment]
`ExecuteScriptCommand` and derived classes LGTM. Can't speak well to the rest though. Was there anything else I should pay attention to?

**Comment by APErebus** [comment]
> `ExecuteScriptCommand` and derived classes LGTM. Can't speak well to the rest though. Was there anything else I should pay attention to?

@jimmyp Not really. I think that was the main thing I wanted your input on. Thanks. I'll work with @LukeButters and the broader SaS team to get the rest ticked off 🙇

File: source/Octopus.Tentacle.Client/Scripts/Models/ExecuteScriptCommand.cs [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+    public class ExecuteScriptCommand
+    {
+        public ExecuteScriptCommand(
+            ScriptTicket scriptTicket,
+            string taskId,
+            string scriptBody,
+            string[] arguments,
+            ScriptIsolationLevel isolationLevel,
+            TimeSpan isolationMutexTimeout,
+            string isolationMutexName,
```
Comment (APErebus) [inline source/Octopus.Tentacle.Client/Scripts/Models/ExecuteScriptCommand.cs:17]: In the process of using this in Server, I found that the mutex name is _always_ set to something, so I changes this to non-nullable.

I also cleaned up these names to be consistent

File: source/Octopus.Tentacle.Client/Scripts/ScriptServiceV3AlphaOrchestrator.cs [hunk 2]
```diff
@@ -46,7 +45,27 @@ class ScriptServiceV3AlphaOrchestrator : ObservingScriptOrchestrator<StartScript
...
         }
 
-        protected override StartScriptCommandV3Alpha Map(StartScriptCommandV3Alpha command) => command;
+        protected override StartScriptCommandV3Alpha Map(ExecuteScriptCommand command)
+        {
+            IScriptExecutionContext executionContext = command switch
+            {
+                ExecuteKubernetesScriptCommand kubernetesScriptCommand => new KubernetesAgentScriptExecutionContext(kubernetesScriptCommand.Image, kubernetesScriptCommand.FeedUrl, kubernetesScriptCommand.FeedUsername, kubernetesScriptCommand.FeedPassword),
+                _ => new LocalShellScriptExecutionContext()
+            };
```
Comment (APErebus) [inline source/Octopus.Tentacle.Client/Scripts/ScriptServiceV3AlphaOrchestrator.cs:54]: The current V3Alpha as the `ScriptExecutionContext` to distinguish, so we look at the command type to create the relevant context

File: source/Octopus.Tentacle.Contracts/Builders/StartScriptCommandV3AlphaBuilder.cs [hunk 3]
```diff
@@ -11,12 +11,12 @@ public class StartScriptCommandV3AlphaBuilder
...
         readonly Dictionary<ScriptType, string> additionalScripts = new();
         StringBuilder scriptBody = new(string.Empty);
-        ScriptIsolationLevel isolation = ScriptIsolationLevel.FullIsolation;
+        ScriptIsolationLevel isolation = ScriptIsolationLevel.NoIsolation;
         TimeSpan scriptIsolationMutexTimeout = TimeSpan.FromMilliseconds(int.MaxValue);
         string scriptIsolationMutexName = "RunningScript";
         string taskId = Guid.NewGuid().ToString();
         ScriptTicket scriptTicket = new UniqueScriptTicketBuilder().Build();
-        TimeSpan? durationStartScriptCanWaitForScriptToFinish = TimeSpan.FromSeconds(5);
+        TimeSpan? durationStartScriptCanWaitForScriptToFinish;
```
Comment (APErebus) [inline source/Octopus.Tentacle.Contracts/Builders/StartScriptCommandV3AlphaBuilder.cs:19]: These were the defaults in the `LatestStartScriptCommandBuilder`, so changed these defaults

File: source/Octopus.Tentacle.Contracts/ScriptServiceV3Alpha/KubernetesAgentScriptExecutionContext.cs [hunk 4]
```diff
@@ -5,7 +5,7 @@ namespace Octopus.Tentacle.Contracts.ScriptServiceV3Alpha
     public class KubernetesAgentScriptExecutionContext : IScriptExecutionContext
     {
         [JsonConstructor]
-        public KubernetesAgentScriptExecutionContext(string image, string? feedUrl, string? feedUsername, string? feedPassword)
+        public KubernetesAgentScriptExecutionContext(string? image, string? feedUrl, string? feedUsername, string? feedPassword)
```
Comment (APErebus) [inline source/Octopus.Tentacle.Contracts/ScriptServiceV3Alpha/KubernetesAgentScriptExecutionContext.cs:8]: technically the JSON value for `image` may not exist, so this should be nullable

File: source/Octopus.Tentacle.Client/Scripts/Models/ExecuteScriptCommand.cs [hunk 5]
```diff
@@ -0,0 +1,54 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using Octopus.Tentacle.Contracts;
+
+namespace Octopus.Tentacle.Client.Scripts.Models
+{
+    public class ExecuteScriptCommand
```
Comment (LukeButters) [inline source/Octopus.Tentacle.Client/Scripts/Models/ExecuteScriptCommand.cs:8]: I wonder if the parent of this and ExecuteKubernetesScriptCommand, ought to be an interface of the minimum things we think any of these commands must have.

This concrete class might become `ExecuteShellScriptCommand`

File: source/Octopus.Tentacle.Client/Scripts/Models/ExecuteScriptCommand.cs (hunk 1 above)
Comment (LukeButters) [reply source/Octopus.Tentacle.Client/Scripts/Models/ExecuteScriptCommand.cs:17]: Nit: The three of those:
```
ScriptIsolationLevel isolationLevel,
            TimeSpan isolationMutexTimeout,
            string isolationMutexName,
```

Could be moved to a single class `ScriptIsolation`

File: source/Octopus.Tentacle.Client/Scripts/Models/ExecuteKubernetesScriptCommand.cs [hunk 6]
```diff
@@ -0,0 +1,36 @@
...
+            string[] arguments,
+            ScriptIsolationLevel isolationLevel,
+            TimeSpan isolationMutexTimeout,
+            string isolationMutexName,
+            Dictionary<ScriptType, string>? additionalScripts,
+            ScriptFile[] additionalFiles,
+            string? image,
+            string? feedUrl,
+            string? feedUsername,
+            string? feedPassword)
```
Comment (LukeButters) [inline source/Octopus.Tentacle.Client/Scripts/Models/ExecuteKubernetesScriptCommand.cs:22]: nit: I wonder if these could be pulled out into some type that groups them

File: source/Octopus.Tentacle.CommonTestUtils/Builders/StartScriptCommandV2BuilderExtensionMethods.cs [hunk 7]
```diff
@@ -15,22 +13,5 @@ public static StartScriptCommandV2Builder WithScriptBodyForCurrentOs(this StartS
 
             return builder;
         }
-
```
Comment (LukeButters) [inline source/Octopus.Tentacle.CommonTestUtils/Builders/StartScriptCommandV2BuilderExtensionMethods.cs:18]: Were these unused?

File: source/Octopus.Tentacle.CommonTestUtils/Builders/StartScriptCommandV3AlphaBuilder.cs [hunk 8]
```diff
@@ -1,22 +1,24 @@
...
+namespace Octopus.Tentacle.CommonTestUtils.Builders
 {
     public class StartScriptCommandV3AlphaBuilder
     {
         readonly List<ScriptFile> files = new();
         readonly List<string> arguments = new();
         readonly Dictionary<ScriptType, string> additionalScripts = new();
         StringBuilder scriptBody = new(string.Empty);
-        ScriptIsolationLevel isolation = ScriptIsolationLevel.FullIsolation;
+        ScriptIsolationLevel isolation = ScriptIsolationLevel.NoIsolation;
```
Comment (LukeButters) [inline source/Octopus.Tentacle.CommonTestUtils/Builders/StartScriptCommandV3AlphaBuilder.cs:16]: I wonder if this should be done in its own PR.

File: source/Octopus.Tentacle.CommonTestUtils/Builders/StartScriptCommandV3AlphaBuilder.cs [hunk 9]
```diff
@@ -1,22 +1,24 @@
...
         TimeSpan scriptIsolationMutexTimeout = TimeSpan.FromMilliseconds(int.MaxValue);
         string scriptIsolationMutexName = "RunningScript";
         string taskId = Guid.NewGuid().ToString();
         ScriptTicket scriptTicket = new UniqueScriptTicketBuilder().Build();
-        TimeSpan? durationStartScriptCanWaitForScriptToFinish = TimeSpan.FromSeconds(5);
+        TimeSpan? durationStartScriptCanWaitForScriptToFinish;
```
Comment (LukeButters) [inline source/Octopus.Tentacle.CommonTestUtils/Builders/StartScriptCommandV3AlphaBuilder.cs:21]: I wonder if this should be done in its own PR.

File: source/Octopus.Tentacle.CommonTestUtils/Builders/StartScriptCommandV3AlphaBuilder.cs [hunk 10]
```diff
@@ -1,22 +1,24 @@
...
 using System.Collections.Generic;
 using System.Text;
+using Octopus.Tentacle.Contracts;
+using Octopus.Tentacle.Contracts.Builders;
 using Octopus.Tentacle.Contracts.ScriptServiceV3Alpha;
 
-namespace Octopus.Tentacle.Contracts.Builders
+namespace Octopus.Tentacle.CommonTestUtils.Builders
 {
     public class StartScriptCommandV3AlphaBuilder
```
Comment (LukeButters) [inline source/Octopus.Tentacle.CommonTestUtils/Builders/StartScriptCommandV3AlphaBuilder.cs:10]: Any change made here will have no impact on customers as no customer ever uses this builder. If that isn't correct this could be an issue as it is changing defaults.

**Review by LukeButters — APPROVED** [review]

File: source/Octopus.Tentacle.CommonTestUtils/Builders/StartScriptCommandV2BuilderExtensionMethods.cs (hunk 7 above)
Comment (APErebus) [reply source/Octopus.Tentacle.CommonTestUtils/Builders/StartScriptCommandV2BuilderExtensionMethods.cs:18]: yes they were

File: source/Octopus.Tentacle.Client/Scripts/Models/ExecuteKubernetesScriptCommand.cs (hunk 6 above)
Comment (APErebus) [reply source/Octopus.Tentacle.Client/Scripts/Models/ExecuteKubernetesScriptCommand.cs:22]: yeah, I can do that 👍

File: source/Octopus.Tentacle.Client/Scripts/Models/ExecuteScriptCommand.cs (hunk 5 above)
Comment (APErebus) [reply source/Octopus.Tentacle.Client/Scripts/Models/ExecuteScriptCommand.cs:8]: yeah, that's a good idea 👍
I might make it an abstract just due to how much shared properties there are, but 👍

File: source/Octopus.Tentacle.CommonTestUtils/Builders/StartScriptCommandV3AlphaBuilder.cs (hunk 10 above)
Comment (APErebus) [reply source/Octopus.Tentacle.CommonTestUtils/Builders/StartScriptCommandV3AlphaBuilder.cs:10]: @LukeButters These builders were only used in Tentacle tests and in Server. In a subsequent PR that consumes these changes, we won't be using those builders in Server, hence why I moved them.

File: source/Octopus.Tentacle.Client/Scripts/Models/ExecuteScriptCommand.cs (hunk 5 above)
Comment (APErebus) [reply source/Octopus.Tentacle.Client/Scripts/Models/ExecuteScriptCommand.cs]: done in [8886e99](https://github.com/OctopusDeploy/OctopusTentacle/pull/845/commits/8886e99b2755537ab48270f1a28141562f3ef682)

File: source/Octopus.Tentacle.Client/Scripts/Models/ExecuteKubernetesScriptCommand.cs (hunk 6 above)
Comment (APErebus) [reply source/Octopus.Tentacle.Client/Scripts/Models/ExecuteKubernetesScriptCommand.cs]: done in [8886e99](https://github.com/OctopusDeploy/OctopusTentacle/pull/845/commits/8886e99b2755537ab48270f1a28141562f3ef682)

File: source/Octopus.Tentacle.CommonTestUtils/Builders/StartScriptCommandV3AlphaBuilder.cs (hunk 9 above)
Comment (APErebus) [reply source/Octopus.Tentacle.CommonTestUtils/Builders/StartScriptCommandV3AlphaBuilder.cs:21]: Done in #855

File: source/Octopus.Tentacle.CommonTestUtils/Builders/StartScriptCommandV3AlphaBuilder.cs (hunk 10 above)
Comment (APErebus) [reply source/Octopus.Tentacle.CommonTestUtils/Builders/StartScriptCommandV3AlphaBuilder.cs:10]: I checked, these were moved into Contracts in #764 in Jan for use in DynamicWorkers, which will be updated to use the new contract type, so they can be moved back.

They are being moved in #855

File: source/Octopus.Tentacle.Tests.Integration/CapabilitiesServiceV2Test.cs [hunk 11]
```diff
@@ -74,7 +74,7 @@ public async Task CapabilitiesResponseShouldBeCached(TentacleConfigurationTestCa
                     .Build())
                 .Build(CancellationToken);
 
-            var startScriptCommand = new LatestStartScriptCommandBuilder()
+            var startScriptCommand = new ExecuteScriptCommandBuilder()
```
[the remark this answers is not in the document]
Comment (APErebus) [reply source/Octopus.Tentacle.Tests.Integration/CapabilitiesServiceV2Test.cs]: Changed in [8886e99](https://github.com/OctopusDeploy/OctopusTentacle/pull/845/commits/8886e99b2755537ab48270f1a28141562f3ef682) to `TestExecuteShellScriptCommandBuilder`

File: source/Octopus.Tentacle.Client/Scripts/Models/ExecuteScriptCommand.cs (hunk 1 above)
Comment (APErebus) [reply source/Octopus.Tentacle.Client/Scripts/Models/ExecuteScriptCommand.cs]: Added in [8886e99](https://github.com/OctopusDeploy/OctopusTentacle/pull/845/commits/8886e99b2755537ab48270f1a28141562f3ef682)

**Comment by APErebus** [comment]
@LukeButters @OctopusDeploy/team-server-at-scale I have updated this PR with the changes discussed. I will be updating my server branch above and doing testing before merging this.

The main changes in [8886e99](https://github.com/OctopusDeploy/OctopusTentacle/pull/845/commits/8886e99b2755537ab48270f1a28141562f3ef682) are to introduce a new `ExecuteShellScriptCommand` and make `ExecuteScriptCommand` abstract. Also introduce new sub-types for the Isolation config & kubernetes image config.

I also copied the builders I'd developed in server into this code so we don't need to duplicate those builders anymore

**Review by LukeButters — APPROVED** [review]

**Comment by APErebus** [comment]
I will be merging these changes when https://github.com/OctopusDeploy/OctopusDeploy/pull/23506 is reviewed and approved

File: source/Octopus.Tentacle.Tests.Integration/ClientScriptExecutionIsolationMutex.cs [hunk 12]
```diff
@@ -32,25 +32,25 @@ public async Task ScriptIsolationMutexFull_EnsuresTwoDifferentScriptsDontRunAtTh
 
             var secondScriptStart = Path.Combine(clientTentacle.TemporaryDirectory.DirectoryPath, "secondScriptStartFile");
 
-            var firstStartScriptCommand = new LatestStartScriptCommandBuilder()
-                .WithScriptBody(new ScriptBuilder()
+            var firstStartScriptCommand = new TestExecuteShellScriptCommandBuilder()
+                .SetScriptBody(new ScriptBuilder()
```
Comment (sburmanoctopus) [inline source/Octopus.Tentacle.Tests.Integration/ClientScriptExecutionIsolationMutex.cs:36]: Hey @APErebus,

I know this is a bit late to the party, but why did we change the convention of using "With" methods in our builders for "Set"?

Comment (APErebus) [reply source/Octopus.Tentacle.Tests.Integration/ClientScriptExecutionIsolationMutex.cs:36]: So I took a bit of inspiration from the Nuke builders and how `Set` feels more specific than `With`. That being said, if this drastically moves away from our standards (is there any doco on this?), I can change them back

Comment (APErebus) [reply source/Octopus.Tentacle.Tests.Integration/ClientScriptExecutionIsolationMutex.cs:36]: Sure, Let me get through these other changes and I'll revert them to `With`
